core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala

```diff
@@ -284,7 +312,60 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
     TestUtils.createTopic(zkUtils, topic, 1, brokerCount, servers, new Properties())
     servers.foreach { server =>
       val logDir = server.logManager.getLog(tp).get.dir.getParent
-      assertEquals(replicaAssignment(new TopicPartitionReplica(topic, 0, server.config.brokerId)), logDir)
+      assertEquals(firstReplicaAssignment(new TopicPartitionReplica(topic, 0, server.config.brokerId)), logDir)
+    }
+
+    // Verify that replica can be moved to the specified log directory after the topic has been created
+    adminClient.alterReplicaLogDirs(secondReplicaAssignment.asJava, new AlterReplicaLogDirsOptions()).all().get()
+    servers.foreach { server =>
+      TestUtils.waitUntilTrue(() => {
+        val logDir = server.logManager.getLog(tp).get.dir.getParent
+        secondReplicaAssignment(new TopicPartitionReplica(topic, 0, server.config.brokerId)) == logDir
+      }, "timed out waiting for replica movement", 6000L)
+    }
+
+    // Verify that replica can be moved to the specified log directory while the producer is sending messages
+    val running = new AtomicBoolean(true)
+    @volatile var numMessages = 0
+    val thread = new Thread() {
+      override def run(): Unit = {
+        val producer = TestUtils.createNewProducer(
+          TestUtils.getBrokerListStrFromServers(servers, protocol = securityProtocol),
+          securityProtocol = securityProtocol,
+          trustStoreFile = trustStoreFile,
+          retries = 0, // Producer should not have to retry when broker is moving replica between log directories.
+          requestTimeoutMs = 2000,
+          acks = -1
+        )
+
+        while (running.get()) {
+          val future = producer.send(new ProducerRecord(topic, s"xxxxxxxxxxxxxxxxxxxx-$numMessages".getBytes))
+          numMessages += 1
+          future.get()
+        }
+        producer.close()
+      }
+    }
+
+    try {
+      thread.start()
+      TestUtils.waitUntilTrue(() => numMessages > 100, "timed out waiting for message produce", 6000L)
+      adminClient.alterReplicaLogDirs(firstReplicaAssignment.asJava, new AlterReplicaLogDirsOptions()).all().get()
+      servers.foreach { server =>
+        TestUtils.waitUntilTrue(() => {
+          val logDir = server.logManager.getLog(tp).get.dir.getParent
+          firstReplicaAssignment(new TopicPartitionReplica(topic, 0, server.config.brokerId)) == logDir
+        }, "timed out waiting for replica movement", 6000L)
+      }
+    } finally {
+      running.set(false)
+      thread.join()
+    }
+
+    // Verify that all messages that are produced can be consumed
+    val consumerRecords = TestUtils.consumeTopicRecords(servers, topic, numMessages, securityProtocol, trustStoreFile)
+    consumerRecords.zipWithIndex.foreach { case (consumerRecord, index) =>
```

Comment:
The reason is that TestUtils.consumeTopicRecords() will throw exception if the ConsumeRecords.size() is not exactly the numMessages(). Do you prefer me to add back assertEquals(numMessages, consumerRecords.size) so that this check is more explicit in the test?

Reply:
Thanks for the explanation. This is fine then.